tensorflow/compiler/tf2xla/functionalize_while.cc

```diff
@@ -200,33 +200,23 @@ Status BuildLoopBody(const Graph& graph, Frame* frame,
     arg_types->push_back(dtype);
 
     TF_ASSIGN_OR_RETURN(Node * arg_node, BuildArgNode(output, dtype, i));
-
-    if (dtype == DT_RESOURCE) {
-      // The convention of the XLA bridge is that resource variable arguments
-      // are only inputs to the loop body and have no corresponding output.
-      // TODO(b/37741920): change the convention so that DT_RESOURCE variables
-      // are both inputs and outputs, and then remove this case.
-      TF_RET_CHECK(arg.is_loop_invariant);
+    TF_ASSIGN_OR_RETURN(Node * retval_node,
+                        BuildRetvalNode(output, dtype, i));
+    if (arg.is_loop_invariant) {
+      // Argument is loop-invariant. Forward it from the Arg to the Retval.
+      // DT_RESOURCE falls into this case.
```

Comment:
I don't understand this comment -- are you saying that `DT_RESOURCE` will always be loop invariant for some check done previously?  If so, maybe we should keep the `TF_RET_CHECK(arg.is_loop_invariant);`?

Reply:
Currently, DT_RESOURCE is always treated as loop invariant AFAIK. Meaning that the is_constant flag of their corresponding Enter node is always set to true.

To make the code more robust to future change, we may add the check as follows. Please let me know what you think.

```suggestion
    if (arg.is_loop_invariant) {
      ...
    } else {
      // DT_RESOURCE should fall into the loop invariant case.
      TF_RET_CHECK(dtype != DT_RESOURCE);
      ...
    }
```

Reply:
> Currently, DT_RESOURCE is always treated as loop invariant AFAIK. Meaning that the is_constant flag of their corresponding Enter node is always set to true.

In other words, the graphs generated by TF today happen to have `DT_RESOURCE` tensors as loop invariant (I would be surprised if this is a semantic property of TF graphs)?  It that's the case then maybe we should be bailing out of the functionalization instead of (or in addition to) `TF_RET_CHECK`ing here?

Reply:
Yes, tf.while always _captures_ resources. I guess nothing can stop you if you truly want to construct a while loop with non-loop-invariant resources by using Enter/Merge/Swtich nodes, etc.

By bailing out, which macro is in your mind? I don't see any ASSERT used under tf2xla (except test codes.) I assume TF_RET_CHECK will also stop the execution if the condition fails?

Reply:
> By bailing out, which macro is in your mind?

I initially had something else in mind, but I think it is probably most appropriate to return a `errors::Unimplemented` in this case (the graph is semantically well formed but we haven't bothered with implementing this corner case because we don't expect to see it in practice).